Allow to specify request fetch mode and cache mode for Wasm target

rhysd opened the pull request:
This patch adds functionality to specify [fetch mode](https://developer.mozilla.org/en-US/docs/Web/API/Request/mode) and [cache mode](https://developer.mozilla.org/en-US/docs/Web/API/Request/cache) to `RequestBuilder`. They are passed to `RequestInit` object of `fetch()` JavaScript function and control the fetching and caching behaviors. These controls are necessary for me since a browser supports CORS and has its own cache logic.

I could not add tests for these implementations since there seem not to be any tests for `wasm32-unknown-unknown` target. If I miss something, please let me know. I'll happily add new test cases for this patch.

rhysd commented on src/wasm/request.rs:
```diff
@@ -3,6 +3,7 @@ use std::fmt;
 
 use http::Method;
 use url::Url;
+pub use web_sys::{RequestMode, RequestCache};
```
Here `web_sys::RequestMode` and `web_sys::RequestCache` are re-exported. I'm not confident to expose `web_sys` things as part of APIs provided by this crate. If we should make our own enum for these modes like below, please let me know. I'll update this patch.

```rust
#[derive(Debug)]
pub enum RequestFetchMode {
    Cors,
    NoCors,
    SameOrigin,
    Navigate,
}
```

seanmonstar replied:
Ya, it's probably best to make the API make sense with and without WASM. Like if reqwest were to eventually add cache support.

Is there a specific mode you need right now, or just wanting to add general support?

thomaseizinger replied:
Alternatively to exposing the actual enums, you could provide specific functions on the builder that use these enums internally (names subject to bike shedding):

```rust
fn fetch_mode_cors(mut self) -> RequestBuilder;
fn fetch_mode_no_cors(mut self) -> RequestBuilder;
fn fetch_mode_same_origin(mut self) -> RequestBuilder;
...
```

This is more verbose but avoids leaking any underlying types (for now).

rhysd replied:
> Is there a specific mode you need right now, or just wanting to add general support?

Currently I need `no-cors` mode. Cache mode is not immediate for me (I just added the API in addition to request fetch mode for general support).

AFAIK CORS is a mechanism only implemented in browser. In terms of aligning APIs with and without Wasm, how about defining general 'request mode' API and defining Wasm-specific variants in it? (it's just an idea, tho)

```rust
#[cfg(not(target_arch = "wasm32"))]
pub enum RequestMode {
    Default,
    // Currently empty, but when some fetch behaviors are implemented as mode,
    // they will go here
}

#[cfg(target_arch = "wasm32")]
pub enum RequestMode {
    Default,
    Cors,
    NoCors,
    SameOrigin,
}
```

seanmonstar replied:
How about let's start conservative and add `fetch_mode_no_cors()`, which in not-WASM just does nothing.

rhysd replied:
It also looks fine for me. I'll update this branch.

rhysd commented:
@seanmonstar @thomaseizinger I changed the implementation at 5a43211 following the discussion in the review. I added minimal API `fetch_mode_no_cors()` to `RequestBuilder`.

seanmonstar commented:
I edited the docs and merged to master in https://github.com/seanmonstar/reqwest/commit/600462378410f85f813f55b42623c6e7f5b06c6e

rhysd commented:
Thank you!
